Add visual bell support

markandrus opened the pull request:
Hi @jwilm,

This project is awesome 👍. I haven't written Rust before, but I wanted to try my hand at a small feature. I think this change is not quite ready, since `draw` is called lazily (so the bell will not be smooth). If you are interested in accepting this feature, perhaps you could share some tips on how to make the right changes? Commit message below.

Thanks,
Mark

***

This commit adds support for a visual bell. Although the Handler in src/ansi.rs warns "Hopefully this is never implemented", I wanted to give it a try. Two new config options are added, with sensible defaults:

* `visual_bell`, which defaults to `true`, and
* `visual_bell_duration`, which defaults to 150 ms.

The visual bell is modeled by VisualBell in src/display.rs. It has a method to ring the bell, `ring`, and another method, `value`. Both return the "value" of the bell, which ramps down from 1.0 to 0.0 at a rate set by `duration`.

jwilm commented:
Hey this looks awesome! Thanks for putting it together.

WRT lazy draw - I think it makes sense to put `VisualBell` on `Term`. Normally after drawing, the dirty flag is cleared. Instead (spitballing), it could be set to `dirty = visual_bell.remaining != 0` or equivalent.

Would need to skip waiting on events if terminal is dirty too.

markandrus commented:
Thanks for the feedback—I will try that out this weekend.

jwilm commented on src/display.rs:
```diff
@@ -88,12 +88,48 @@ impl From<renderer::Error> for Error {
+                let duration = now.duration_since(earlier);
+                let elapsed_ms = duration.as_secs() as f64 * 1_000f64 +
+                                 duration.subsec_nanos() as f64 / 1_000_000f64;
+                (1.0 - elapsed_ms / self.duration as f64).max(0.0)
```
Linear is definitely one way to go, but I think it might be good to consider some sort of "ease out" function where  most of the intensity disappears quickly. A simple example of this would be squaring or cubing the normalized elapsed time.

jwilm commented on src/display.rs:
```diff
@@ -88,12 +88,48 @@ impl From<renderer::Error> for Error {
 
+pub struct VisualBell {
+    /// Visual bell duration
+    duration: u16,
```
Seems like we would want to store this as a `std::time::Duration`.

jwilm commented on src/term/mod.rs:
```diff
@@ -345,6 +345,8 @@ pub struct Term {
 
     pub dirty: bool,
 
+    pub rang_bell: bool,
```
I _think_ this flag will become obsolete once the VisualBell is moved onto this type.

jwilm commented on src/display.rs:
```diff
@@ -88,12 +88,48 @@ impl From<renderer::Error> for Error {
+    duration: u16,
+
+    /// The last time the visual bell rang, if at all
+    last_rang: Option<::std::time::Instant>,
```
Bikeshed - maybe calling this `start` or `start_time` would be more appropriate?

jwilm commented on src/config.rs:
```diff
@@ -249,6 +249,14 @@ pub struct Config {
+
+    /// Visual bell duration in milliseconds
+    #[serde(default="default_visual_bell_duration")]
+    visual_bell_duration: u16
```
The config for this would be simplified by just having a `visual_bell_duration: Option<u16>`. The `visual_bell` flag would no longer be needed.

markandrus commented:
Hi @jwilm,

Mind taking another look? I believe I've addressed all of your feedback. Can you please take a look at "Configure whether the Processor waits for events" in particular? I'm not sure if this approach is correct (namely, introducing a `wait_for_event` property to the Processor, set by `terminal.dirty`). I also ran with your suggestion of an easing function, and introduced perhaps more configurability than anyone will want or need :-D but the defaults ought to be sensible...

I simplified the config so that (in addition to omitting the `visual_bell` property entirely), you can either specify only the `duration`, e.g.

```yaml
visual_bell: 150
```

or, specify the `duration`, `animation`, and any future options:

```yaml
visual_bell:
  duration: 150
  animation: EaseOutExpo
```

jwilm commented on src/config.rs:
```diff
@@ -173,6 +174,110 @@ impl IndexMut<usize> for ColorList {
+            formatter.write_str("u16 or map")
+        }
+
+        fn visit_i64<E>(self, value: i64) -> ::std::result::Result<T, E>
```
Why not implement visit_u64 instead? That would at least make sure a negative value cannot be passed.

markandrus replied:
Removed the config overloading along with this code.

jwilm commented:
Forgot to reply to your comments -

> Can you please take a look at "Configure whether the Processor waits for events" in particular? I'm not sure if this approach is correct (namely, introducing a wait_for_event property to the Processor, set by terminal.dirty).

Seemed fine to me

> I also ran with your suggestion of an easing function, and introduced perhaps more configurability than anyone will want or need :-D but the defaults ought to be sensible...

Yeah, this is great. I think some documentation in the default config file would be good so people know what options are available. The default is good.

> I simplified the config so that (in addition to omitting the visual_bell property entirely), you can either specify only the duration.

Seems nice. It's a bummer that so much code is needed for it, though. I would almost be inclined to require the map form just to keep the LOC for this feature down.

markandrus commented on src/term/mod.rs:
```diff
@@ -299,6 +300,101 @@ pub struct Cursor {
+    /// duration.
+    pub fn intensity_at_instant(&self, instant: Instant) -> f64 {
+        let x = if self.duration == Duration::from_secs(0) {
+            1.0
```
Yeah, I can see how this reads strangely. I've refactored it a bit, renamed some variables, and added comments. Hopefully it will be clearer.

markandrus commented:
> It appears that `VisualBell.start_time` is never reset to `None`. Everything seems to work fine without it, but it's confusing semantically.

I think it made a little more since when the property was called `last_rang`. I specifically wanted to model "the last time the bell had been rung (if at all)", so that `intensity_at_instant` could return 0.0 in the "never rung" case. Once the bell has been rung, computing `time` is always the same (ratio of time elapsed since `start_time` to `duration`, capped at 1.0), so we don't strictly need to reset `start_time`.

I did look into how we could do this, e.g. in `completed`, although that would require converting `&self` to `&mut self`, which feels wrong. The other complication I saw is that we'd probably want to reset `start_time` once `duration` has elapsed, but any method we might define to do so would be called lazily I think... perhaps documenting around this would be better? WDYT?

> Oh, and then I was also confused about why `intensity_at_instant` defaults to 1.0 instead of 0.0. This seems backwards to me.

I see what you mean. I've refactored—hopefully it reads more clearly.

> I think some documentation in the default config file would be good so people know what options are available.

Updated 👍 

> Seems nice. It's a bummer that so much code is needed for it, though. I would almost be inclined to require the map form just to keep the LOC for this feature down.

Yeah, I wish I could've just set the property to `Either<u16, VisualBellConfig>` and gotten the deserialization for free. Anyway, I've removed it.

jwilm commented on src/event.rs:
```diff
@@ -250,34 +253,42 @@ impl<N: Notify> Processor<N> {
             }
         }
 
+        self.wait_for_event = !terminal.dirty;
```
This seems like a weird place to set `wait_for_event`. It's possible that the terminal would be undirtied by the next draw and the correct action would be to wait on the next iteration.

I would suggest that we pass `!terminal.bell.completed()` (guessing at names) to the event processing for the value of `wait_for_event` (ie, at the `process_events` call site).

In replace of this line, we would want to update `terminal.dirty ||= !terminal.bell.completed()`.

markandrus replied:
I'm struggling to make this change—I can't compute `wait_for_event` from within the `process_events`, because it depends on `terminal` which isn't initialized until after calling `window.wait_events().next()` (itself gated by `wait_for_event`).

So I've also looked at passing in `wait_for_event` as an argument to `process_events` in `src/main.rs`. Each time in the loop, I could update `wait_for_event` after `draw` based on `completed`, but then I hit a compile error because I think `draw` takes ownership of the `terminal`? Anyway, before I go further, does it sound like I am on the right track?

jwilm replied:
hmm, maybe we just scratch this change for now.

markandrus replied:
OK, well I am open to the change, or we could capture the desired change in an issue... I will take another look tonight and see if I can find a solution that makes sense.

jwilm replied:
I think we can just put it in an issue. The issue is that 1 extra frame will be drawn whether it's needed or not. The locking of `Term` is going to be annoying to resolve it.

We should probably try and merge this branch and file an issue as you suggested.

jwilm commented on src/event.rs:
```diff
@@ -250,34 +253,42 @@ impl<N: Notify> Processor<N> {
+                mouse_bindings: &self.mouse_bindings[..]
+            };
+
+            match event {
```
This could be simplified to

```rust
if let Some(event) = event {
    process!(event);
}
```

jwilm commented on src/display.rs:
```diff
@@ -270,7 +270,7 @@ impl Display {
     pub fn draw(&mut self, mut terminal: MutexGuard<Term>, config: &Config, selection: &Selection) {
         // Clear dirty flag
-        terminal.dirty = false;
+        terminal.dirty = !terminal.visual_bell.completed();
```
This change becomes unnecessary with the suggested `wait_event` changes.

jwilm commented:
Assuming you're happy with all of this, it's ready to go from my perspective. Would you mind squashing and rebasing?

markandrus commented:
Thanks @jwilm—I have squashed and rebased.

jwilm commented:
Landed in fbc7b7227171b41d96ca52df52e4cf1833f5fc6f

jwilm commented:
Thanks again for all the work on this! I'll be using it. :smile:
